src/nvim/ex_getln.c

```diff
@@ -1780,24 +1780,29 @@ getexmodeline (
         msg_col = startcol;
         msg_clr_eos();
         line_ga.ga_len = 0;
-        continue;
+        goto redraw;
       }
 
+      int num_spaces;
       if (c1 == Ctrl_T) {
         int sw = get_sw_value(curbuf);
 
         p = (char_u *)line_ga.ga_data;
         p[line_ga.ga_len] = NUL;
         indent = get_indent_str(p, 8, FALSE);
-        indent += sw - indent % sw;
+        num_spaces = sw - indent % sw;
+        indent += num_spaces;
 add_indent:
-        while (get_indent_str(p, 8, FALSE) < indent) {
+        if (num_spaces > 0) {
+          ga_grow(&line_ga, num_spaces + 1);
+          p = (char_u *)line_ga.ga_data;
           char_u *s = skipwhite(p);
 
-          ga_grow(&line_ga, 1);
-          memmove(s + 1, s, line_ga.ga_len - (s - p) + 1);
-          *s = ' ';
-          ++line_ga.ga_len;
+          // Insert spaces after leading whitespaces.
+          memmove(s + num_spaces, s, line_ga.ga_len - (s - p) + 1);
+          memset(s, ' ', num_spaces);
```

Comment:
We first need to make room for the new spaces between the leading whitespaces and the rest of the string by shifting to the right. Otherwise applying `memset()` before the `memmove()` would overwrite characters.